---

# Del Texto a Matrices
### ¿Por qué la máquina puede interpretar?



---
---
## 🧮 I. Document Feature Matrix (dfm)

### 🧠 ¿Qué es una DFM en NLP?

Para que una computadora pueda analizar texto, primero es necesario representar los documentos en una forma numérica. Una de las representaciones más utilizadas en el NLP tradicional es la **Document-Feature Matrix (DFM)**.

Una **DFM (Document-Feature Matrix)** es una representación matricial de un corpus textual donde:

- **Cada fila** representa un documento.
- **Cada columna** representa una característica (*feature*) del texto, normalmente una palabra, aunque también puede ser un n-grama u otra unidad lingüística.
- **Cada celda** indica la frecuencia con que esa característica aparece en ese documento.

---

#### 📊 Ejemplo simple

Corpus:

- Doc1: "el gato duerme"
- Doc2: "el perro duerme"

Matriz DFM:

| Documento | el | gato | perro | duerme |
|-----------|----|------|--------|--------|
| Doc1      | 1  | 1    | 0      | 1      |
| Doc2      | 1  | 0    | 1      | 1      |

---

#### 📌 ¿Por qué es importante?

La DFM convierte un conjunto de documentos en una estructura numérica sobre la cual pueden aplicarse numerosos métodos estadísticos y de aprendizaje automático, entre ellos:

- Clasificación de texto.
- Modelamiento de tópicos (LDA, STM).
- Clustering o segmentación.
- Análisis de similitud entre documentos.
- Regresiones y otros modelos estadísticos basados en texto.

La DFM constituye la representación estándar del **NLP tradicional** y posee una gran ventaja: es altamente interpretable, ya que cada columna corresponde a un término cuyo significado conocemos.

---

#### 💡 Una limitación de la DFM

La DFM representa un documento únicamente mediante la frecuencia de sus palabras. Esto significa que dos palabras con significados similares, pero escritas de forma diferente, se consideran características completamente distintas.

Por ejemplo, las palabras:

- automóvil
- coche
- vehículo

aparecen como tres columnas independientes en una DFM, aunque semánticamente estén muy relacionadas.

Esta es una de las principales limitaciones de las representaciones basadas en frecuencias.

---

#### 🤖 ¿Se utiliza la DFM en los LLM?

**No como representación principal.**

Los modelos modernos basados en **Transformers** (como BERT, GPT, Llama o Mistral) no representan los documentos mediante una DFM. En su lugar utilizan representaciones numéricas que incorporan información sobre el contexto y el significado de las palabras.

Más adelante estudiaremos estas representaciones y veremos cómo permiten capturar relaciones semánticas que una DFM no puede representar.

---

#### 📊 ¿Cuándo utilizar una DFM?

| Escenario | ¿Usar DFM? |
|-----------|------------|
| Modelos tradicionales (SVM, Naive Bayes, Regresión Logística) | ✅ Sí |
| Modelamiento de tópicos (LDA, STM) | ✅ Sí |
| Cálculo de TF-IDF | ✅ Sí |
| Análisis exploratorio y visualización textual | ✅ Sí |
| Transformers y Large Language Models (LLMs) | ❌ No como representación principal |

---

> 📌 **Idea clave:** La DFM sigue siendo una herramienta fundamental del NLP tradicional porque es simple, interpretable y muy eficiente. Sin embargo, representa los documentos únicamente mediante la frecuencia de sus términos. En las próximas unidades estudiaremos otras formas de representar texto que permiten capturar relaciones semánticas entre palabras y documentos.

In [1]:
#install.packages("quanteda")
suppressPackageStartupMessages(library(quanteda))


### ✂️ Tokenización: el paso previo a la DFM

Antes de construir una **Document-Feature Matrix (DFM)**, es necesario transformar el corpus en una secuencia de **tokens**. Cada token representa una unidad del texto (normalmente una palabra) que posteriormente podrá convertirse en una característica (*feature*) de la matriz.

En esta etapa también es habitual aplicar algunas operaciones de preprocesamiento para reducir la variabilidad del texto, tales como:

- **Eliminar signos de puntuación**.
- **Eliminar *stopwords***.
- **Convertir todas las palabras a minúsculas**.

Estas transformaciones permiten obtener una representación más limpia y consistente del corpus antes de construir la DFM.

Por ejemplo, utilizando el corpus `data_corpus_inaugural`, podemos realizar este proceso mediante:

```r
toks_nopunct_stop_low <- tokens(data_corpus_inaugural, remove_punct = TRUE) %>%
  tokens_remove(pattern = stopwords("en")) %>%
  tokens_tolower()
```

> 📌 En el NLP tradicional, la calidad de la DFM depende en gran medida de las decisiones tomadas durante el preprocesamiento del texto.

In [2]:
toks_nopunct_stop_low <- tokens(data_corpus_inaugural, remove_punct = TRUE) %>%
                         tokens_remove(pattern = stopwords("en")) %>%
                         tokens_tolower()
toks_nopunct_stop_low

Tokens consisting of 60 documents and 4 docvars.
1789-Washington :
 [1] "fellow-citizens" "senate"          "house"           "representatives"
 [5] "among"           "vicissitudes"    "incident"        "life"           
 [9] "event"           "filled"          "greater"         "anxieties"      
[ ... and 640 more ]

1793-Washington :
 [1] "fellow"     "citizens"   "called"     "upon"       "voice"     
 [6] "country"    "execute"    "functions"  "chief"      "magistrate"
[11] "occasion"   "proper"    
[ ... and 50 more ]

1797-Adams :
 [1] "first"       "perceived"   "early"       "times"       "middle"     
 [6] "course"      "america"     "remained"    "unlimited"   "submission" 
[11] "foreign"     "legislature"
[ ... and 1,058 more ]

1801-Jefferson :
 [1] "friends"   "fellow"    "citizens"  "called"    "upon"      "undertake"
 [7] "duties"    "first"     "executive" "office"    "country"   "avail"    
[ ... and 801 more ]

1805-Jefferson :
 [1] "proceeding"    "fellow"        "ci

---

### 🏗️ Crear una DFM

Una vez tokenizado el corpus, podemos construir la **Document-Feature Matrix (DFM)**. En `quanteda`, esto se realiza con la función `dfm()`.

La DFM resume cuántas veces aparece cada *feature* en cada documento del corpus.



In [3]:
dfm_inaug_nostem <- dfm(toks_nopunct_stop_low)
print(dfm_inaug_nostem)


Document-feature matrix of: 60 documents, 9,435 features (92.79% sparse) and 4 docvars.
                 features
docs              fellow-citizens senate house representatives among
  1789-Washington               1      1     2               2     1
  1793-Washington               0      0     0               0     0
  1797-Adams                    3      1     0               2     4
  1801-Jefferson                2      0     0               0     1
  1805-Jefferson                0      0     0               0     7
  1809-Madison                  1      0     0               0     0
                 features
docs              vicissitudes incident life event filled
  1789-Washington            1        1    1     2      1
  1793-Washington            0        0    0     0      0
  1797-Adams                 0        0    2     0      0
  1801-Jefferson             0        0    1     0      0
  1805-Jefferson             0        0    2     0      0
  1809-Madison               


---
### 🧩 ¿Qué es un *Feature* en una Document-Feature Matrix (DFM)?

En una **Document-Feature Matrix (DFM)**, cada fila representa un documento del corpus y cada **columna representa una característica (*feature*)** del texto.

Desde un punto de vista estadístico, un **feature puede interpretarse como una variable**. Cada documento constituye una observación y cada feature corresponde a una variable cuyo valor indica cuántas veces aparece ese término en el documento.

En nuestro ejemplo, la DFM contiene **9.438 features**, es decir, **9.438 variables** que describen el contenido textual del corpus.

---

#### 📌 ¿Qué puede ser un *feature*?

Dependiendo de cómo se construya la DFM, un *feature* puede corresponder a:

- Una **palabra**, como `"freedom"`, `"government"` o `"peace"`.
- Un **número**, como `"1776"`.
- Un **n-grama**, como `"united states"` o `"human rights"`.
- Un símbolo o signo, si éstos no fueron eliminados durante el preprocesamiento.

---

#### 📊 ¿Qué representan los *features*?

Los *features* constituyen la información utilizada para representar cuantitativamente cada documento y sirven como variables de entrada para numerosos métodos de análisis de texto, entre ellos:

- Análisis de frecuencias.
- Clasificación de documentos.
- Modelamiento de tópicos (LDA, STM).
- Análisis de similitud.
- Clustering.
- Regresiones y otros modelos estadísticos que incorporan información textual.

En otras palabras, la DFM transforma un conjunto de documentos en una **matriz de datos**, donde las observaciones son documentos y las variables son los *features* extraídos del texto.

---

#### 💡 Una observación importante

En una DFM tradicional, cada *feature* se considera una variable independiente. Por ejemplo, las palabras:

- automóvil
- coche
- vehículo

se almacenan como tres columnas distintas, aunque tengan significados muy similares.

Esta característica hace que la DFM sea muy interpretable, pero también explica por qué los enfoques modernos buscan representaciones capaces de capturar relaciones semánticas entre los términos.

---

### 🧠 ¿Qué ocurre con el número de *features* si aplicamos *stemming*?

Cuando aplicamos **stemming**, el número de **features** (columnas) de la DFM suele **disminuir**, ya que distintas formas de una misma palabra se agrupan bajo una única raíz (*stem*).

---

#### 📉 ¿Por qué se reduce?

El **stemming** reemplaza diferentes variantes morfológicas de una palabra por una representación común. Por ejemplo:

| Formas originales | Stem resultante |
|------------------|-----------------|
| correr, corriendo, corrió | corr |
| hablar, habló, hablando | habl |
| feliz, felices | feliz |

En lugar de crear una columna para cada forma de la palabra, la DFM almacena una única columna correspondiente al *stem*.

---

#### 📊 ¿Qué efecto tiene sobre la DFM?

Como varias palabras pasan a representarse mediante un mismo *feature*:

- Disminuye el número de columnas de la matriz.
- Aumenta la frecuencia de los *stems* resultantes.
- Se reduce la **dimensionalidad** de la representación.
- Los modelos tradicionales pueden generalizar mejor cuando distintas formas de una palabra expresan una misma idea.

---

#### 🧪 Ejemplo

**Sin stemming:**

| Documento | hablar | habló | hablando |
|-----------|--------|-------|----------|
| Doc1 | 0 | 1 | 1 |

**Con stemming:**

| Documento | habl |
|-----------|------|
| Doc1 | 2 |

---

#### ⚖️ Ventajas y desventajas

##### ✅ Ventajas

- Reduce la dimensionalidad de la DFM.
- Disminuye el ruido morfológico.
- Agrupa variantes de una misma palabra.
- Puede mejorar el desempeño de modelos basados en frecuencias.

##### ⚠️ Desventajas

- Se pierde información gramatical.
- Algunas palabras distintas pueden compartir el mismo *stem*.
- El *stem* obtenido no siempre corresponde a una palabra existente.

---

#### 🤖 ¿Es necesario aplicar *stemming* en los LLM?

Generalmente **no**.

Los modelos modernos basados en **Transformers** utilizan tokenización mediante subpalabras y representaciones contextuales, por lo que normalmente no necesitan reducir explícitamente las palabras a una raíz común.

Mientras que el *stemming* busca disminuir la dimensionalidad de una DFM, los LLM aprenden automáticamente relaciones entre palabras como:

- correr
- corriendo
- corrió

sin necesidad de transformarlas previamente.

---

> 📌 **Idea clave:** El *stemming* es una técnica muy útil cuando representamos texto mediante una DFM. Sin embargo, en los modelos modernos basados en Transformers suele ser innecesario, ya que estos aprenden relaciones entre las distintas formas de una palabra durante su entrenamiento.

Apliquemos stemming:

In [4]:
toks_nopunct_stop_low_stem <- tokens_wordstem(toks_nopunct_stop_low)
dfm_inaug_stem <- dfm(toks_nopunct_stop_low_stem)
print(dfm_inaug_stem)

Document-feature matrix of: 60 documents, 5,553 features (89.47% sparse) and 4 docvars.
                 features
docs              fellow-citizen senat hous repres among vicissitud incid life
  1789-Washington              1     1    2      2     1          1     1    1
  1793-Washington              0     0    0      0     0          0     0    0
  1797-Adams                   3     1    3      3     4          0     0    2
  1801-Jefferson               2     0    0      1     1          0     0    1
  1805-Jefferson               0     0    0      0     7          0     0    2
  1809-Madison                 1     0    0      1     0          1     0    1
                 features
docs              event fill
  1789-Washington     2    1
  1793-Washington     0    0
  1797-Adams          0    0
  1801-Jefferson      0    0
  1805-Jefferson      1    0
  1809-Madison        0    1
[ reached max_ndoc ... 54 more documents, reached max_nfeat ... 5,543 more features ]


---

### 🧮 Sparsity

```text
0 0 0 1 0
0 0 0 0 0
1 0 0 0 0
0 0 2 0 0
```

Al construir una **Document-Feature Matrix (DFM)** es habitual obtener una salida como:

> *Document-feature matrix of: 60 documents, 5,556 features (89.47% sparse) and 4 docvars.*

El concepto de **sparsity** (dispersión o escasez) es una de las características fundamentales de las representaciones tradicionales del texto.

---

#### 🔍 ¿Qué significa *sparsity*?

La **sparsity** corresponde al **porcentaje de valores iguales a cero** en una matriz.

En una DFM, esto ocurre porque cada documento contiene sólo una pequeña fracción de todas las palabras presentes en el corpus.

Por ejemplo, una sparsity de **89.47%** significa que aproximadamente **9 de cada 10 celdas de la matriz contienen un cero**, ya que la mayoría de las palabras no aparecen en la mayoría de los documentos.

---

#### 📌 ¿Por qué ocurre?

Supongamos que nuestro vocabulario contiene **5.556 palabras**.

Cada documento, sin embargo, utiliza sólo unas pocas centenas de ellas.

Como consecuencia, la mayor parte de la matriz queda formada por ceros.

Esta es una propiedad natural de las representaciones basadas en frecuencias de palabras.

---

### ⚖️ ¿Qué implicancias tiene?

Una matriz muy dispersa presenta varias ventajas y desafíos.

#### ✅ Ventajas

- Es altamente interpretable.
- Conserva la frecuencia exacta de cada término.
- Funciona muy bien en numerosos métodos clásicos de NLP.

#### ⚠️ Desafíos

- Aumenta el consumo de memoria.
- Incrementa el costo computacional.
- Puede dificultar el entrenamiento de algunos modelos.
- No representa relaciones semánticas entre palabras.

---

### 🛠️ ¿Cómo reducir la sparsity?

Algunas estrategias habituales son:

- Aplicar **stemming** o lematización.
- Eliminar términos muy poco frecuentes mediante `dfm_trim()`.
- Utilizar ponderaciones como **TF-IDF**.
- Reducir la dimensionalidad cuando sea apropiado.

---

### 💡 Una mirada hacia adelante

La alta sparsity es una de las razones por las que surgieron nuevas formas de representar texto.

En las últimas unidades del curso utilizaremos representaciones modernas que ya no describen un documento mediante miles de conteos de palabras, sino mediante un conjunto mucho más compacto de valores numéricos capaces de capturar relaciones semánticas entre palabras y documentos.

> 📌 La DFM es una **representación dispersa (sparse)** del texto. Más adelante utilizaremos representaciones **densas (dense)** que constituyen la base de los modelos modernos basados en Transformers y Large Language Models (LLMs).

---
---

## 🎯 II. Selección de *Features* o Características

Una vez que los documentos han sido representados mediante una **Document-Feature Matrix (DFM)**, el siguiente paso consiste en **explorar y seleccionar las características (*features*)** que serán utilizadas en el análisis.

Las *features* corresponden normalmente a **palabras**, aunque también pueden ser **n-gramas**, números u otras unidades lingüísticas, dependiendo de cómo se haya construido la DFM.

Una buena selección de *features* permite construir modelos más simples, eficientes e interpretables.

---

#### 🔍 Funciones útiles en `quanteda`

- `topfeatures(dfm)`: muestra las características más frecuentes del corpus.
- `dfm_trim()`: elimina *features* según su frecuencia o el número de documentos en que aparecen.
- `dfm_select()`: selecciona o elimina términos específicos utilizando patrones.
- `featnames(dfm)`: devuelve el nombre de todas las *features* de la matriz.
- `docfreq(dfm)`: indica en cuántos documentos aparece cada *feature*.

---

#### 💡 ¿Por qué es importante seleccionar adecuadamente las *features*?

Una buena selección de características permite:

- Reducir el número de *features* y, por tanto, el costo computacional.
- Eliminar términos poco informativos o muy poco frecuentes.
- Disminuir el ruido presente en la representación del texto.
- Facilitar la interpretación de los resultados.
- Reducir el riesgo de sobreajuste (*overfitting*) en modelos supervisados.

Estas ventajas son especialmente importantes antes de aplicar técnicas como clasificación de documentos, análisis de sentimientos o modelamiento de tópicos.

---

> 📌 **Idea clave:** En NLP tradicional, la calidad de un modelo depende en gran medida de la calidad de las *features* utilizadas. Una buena selección de características suele traducirse en modelos más eficientes, más interpretables y con mejor capacidad de generalización.

---

### **topfeatures()**

La función `topfeatures()` permite identificar las características (palabras o términos) más frecuentes dentro de una matriz documento-característica (*dfm*). Este tipo de inspección es fundamental en etapas tempranas del análisis textual, ya que permite:

- Observar qué términos dominan el corpus.
- Detectar posibles *stopwords* residuales.
- Guiar la selección o exclusión de términos para análisis posteriores.

Por ejemplo, aplicar `topfeatures(dfmat, n = 10)` muestra las 10 palabras más frecuentes en el objeto `dfmat`.

Esta función es especialmente útil antes de aplicar modelos como análisis de sentimientos, clasificación o modelado de tópicos, ya que ayuda a entender la estructura léxica general del corpus.



In [5]:
print(topfeatures(dfm_inaug_nostem))

    people government         us        can       must       upon      great 
       592        575        507        489        377        371        354 
       may     states      world 
       343        343        329 


In [6]:
print(topfeatures(dfm_inaug_stem))

 nation  govern   peopl      us     can   state   great   power    must countri 
    714     668     640     507     489     463     389     384     377     376 


¿Que diferencias se notan entre las *topfeatures* de un documento stemmed y el documento not-stemmed?


Por ejemplo, antes del stemming, "nation" ni siquiera aparecía entre las 10 primeras características, lo que ilustra cómo el stemming permite captar el uso de palabras similares de forma diferente.

In [7]:
print(topfeatures(dfm_inaug_nostem,20))

    people government         us        can       must       upon      great 
       592        575        507        489        377        371        354 
       may     states      world     nation    country      shall      every 
       343        343        329        324        323        316        309 
       one      peace        new      power        now     public 
       274        259        256        246        238        227 


¿Por qué pasa esto? ¿Se le ocurre como investigarlo?

In [8]:
# Key words in context
kw_nation <- kwic(toks_nopunct_stop_low, pattern = "nation*")
print(head(kw_nation, 10))

Keyword-in-context with 10 matches.                                                                             
 [1789-Washington, 159]           almighty rules universe presides councils |
 [1789-Washington, 225]           every step advanced character independent |
 [1789-Washington, 361] assemblage communities interests another foundation |
 [1789-Washington, 423]                    smiles heaven can never expected |
       [1797-Adams, 56]    signally protected country first representatives |
      [1797-Adams, 169]             faith loss consideration credit foreign |
      [1797-Adams, 179]  partial conventions insurrection threatening great |
      [1797-Adams, 262]        adapted genius character situation relations |
      [1797-Adams, 360]               upon peace order prosperity happiness |
      [1797-Adams, 381]               ancient idea congregations men cities |
                                                               
 nations  | whose providential aids can sup

In [9]:
?topfeatures

topfeatures              package:quanteda              R Documentation

_I_d_e_n_t_i_f_y _t_h_e _m_o_s_t _f_r_e_q_u_e_n_t _f_e_a_t_u_r_e_s _i_n _a _d_f_m

_D_e_s_c_r_i_p_t_i_o_n:

     List the most (or least) frequently occurring features in a dfm,
     either as a whole or separated by document.

_U_s_a_g_e:

     topfeatures(
       x,
       n = 10,
       decreasing = TRUE,
       scheme = c("count", "docfreq"),
       groups = NULL
     )
     
_A_r_g_u_m_e_n_t_s:

       x: the object whose features will be returned

       n: how many top features should be returned

decreasing: If ‘TRUE’, return the ‘n’ most frequent features; otherwise
          return the ‘n’ least frequent features

  scheme: one of ‘count’ for total feature frequency (within ‘group’ if
          applicable), or ‘docfreq’ for the document frequencies of
          features

  groups: grouping variable for sampling, equal in length to the number
  

----

#### **Agrupando por alguna variable (docvar)**

In [10]:
head(docvars(data_corpus_inaugural))

,Year,President,FirstName,Party
,<int>,<chr>,<chr>,<fct>
1,1789,Washington,George,none
2,1793,Washington,George,none
3,1797,Adams,John,Federalist
4,1801,Jefferson,Thomas,Democratic-Republican
5,1805,Jefferson,Thomas,Democratic-Republican
6,1809,Madison,James,Democratic-Republican


In [11]:
print(head(topfeatures(dfm_inaug_nostem,5, groups = President)))

$Adams
government     people      union       upon    country 
        33         27         22         21         18 

$Biden
     us america     can     one  nation 
     27      18      16      15      12 

$Buchanan
      states        shall constitution          may       people 
          22           18           17           15           13 

$Bush
freedom  nation      us america     can 
     36      27      27      27      24 

$Carter
   can nation    new   must     us 
    13     10      9      8      8 

$Cleveland
    people government     public      every      shall 
        35         29         19         14         14 



In [12]:
print(head(topfeatures(dfm_inaug_nostem,5, groups = Party)))

$Democratic
        us     people        can government       must 
       222        199        173        143        138 

$`Democratic-Republican`
government      great     states        war        may 
        68         61         56         51         49 

$Federalist
    people government        may    nations    country 
        20         16         13         11          9 

$none
       can      every government        may    present 
         9          9          9          7          6 

$Republican
    people government        can         us       must 
       272        251        230        220        202 

$Whig
  government       states       people        power constitution 
          88           61           57           57           55 



In [13]:
foo <- topfeatures(dfm_inaug_nostem,5, groups = Party)
print(foo)
class(foo)
str(foo)

$Democratic
        us     people        can government       must 
       222        199        173        143        138 

$`Democratic-Republican`
government      great     states        war        may 
        68         61         56         51         49 

$Federalist
    people government        may    nations    country 
        20         16         13         11          9 

$none
       can      every government        may    present 
         9          9          9          7          6 

$Republican
    people government        can         us       must 
       272        251        230        220        202 

$Whig
  government       states       people        power constitution 
          88           61           57           57           55 



[1] "list"

List of 6
 $ Democratic           : Named num [1:5] 222 199 173 143 138
  ..- attr(*, "names")= chr [1:5] "us" "people" "can" "government" ...
 $ Democratic-Republican: Named num [1:5] 68 61 56 51 49
  ..- attr(*, "names")= chr [1:5] "government" "great" "states" "war" ...
 $ Federalist           : Named num [1:5] 20 16 13 11 9
  ..- attr(*, "names")= chr [1:5] "people" "government" "may" "nations" ...
 $ none                 : Named num [1:5] 9 9 9 7 6
  ..- attr(*, "names")= chr [1:5] "can" "every" "government" "may" ...
 $ Republican           : Named num [1:5] 272 251 230 220 202
  ..- attr(*, "names")= chr [1:5] "people" "government" "can" "us" ...
 $ Whig                 : Named num [1:5] 88 61 57 57 55
  ..- attr(*, "names")= chr [1:5] "government" "states" "people" "power" ...


### Esquemas: `scheme = "count"` vs `scheme = "docfreq"`

En funciones como `topfeatures()` o `dfm_weight()` del paquete `quanteda`, el argumento `scheme` permite especificar cómo se calculará la importancia o el peso de las *features* (palabras o términos) dentro del corpus.

#### 🔹 `scheme = "count"`

- Calcula la **frecuencia total** de cada término en todo el corpus.
- Es decir, cuenta cuántas veces aparece cada palabra en todos los documentos.
- Este esquema resalta términos que son frecuentes en general, aunque no necesariamente representativos de un documento en particular.

✅ Útil para:
- Identificar las palabras más comunes en el corpus.
- Crear nubes de palabras (*wordclouds*) generales.

---

#### 🔸 `scheme = "docfreq"`

- Calcula la **frecuencia documental**: el número de documentos en los que aparece al menos una vez una palabra.
- No cuenta cuántas veces aparece, sino **en cuántos documentos distintos** aparece.

✅ Útil para:
- Identificar términos que están distribuidos en muchos documentos.
- Evitar sesgos por documentos largos o repetitivos.
- Preprocesamiento para aplicar métodos como **tf-idf** o modelamiento de tópicos.

> ⚠️ Ambos esquemas capturan diferentes aspectos del uso de las palabras en el corpus. Su elección depende del objetivo del análisis: frecuencia total vs. presencia distribuida.


In [14]:
print(topfeatures(dfm_inaug_nostem,5, scheme = "count"))


    people government         us        can       must 
       592        575        507        489        377 


In [15]:
print(topfeatures(dfm_inaug_nostem,5, scheme = "docfreq"))

 people     can   great      us country 
     58      57      57      57      55 


Esto significa que "*poeple*" aparece 592 veces en 58 de los 60 documentos totales del corpus.



---

### **dfm_select()**, **dfm_remove()**, **dfm_trim()**

Una vez creada la matriz de características (dfm), `quanteda` ofrece funciones clave para refinarla seleccionando o eliminando términos según diversos criterios: 

- **`dfm_select()`**: Permite **seleccionar** características específicas del `dfm`, utilizando patrones exactos, expresiones regulares o listas. Es útil para centrar el análisis en términos clave.

- **`dfm_remove()`**: Realiza lo opuesto a `dfm_select()`, eliminando características específicas (como *stopwords*, números, u otras palabras irrelevantes).

- **`dfm_trim()`**: Reduce la dimensionalidad eliminando características **poco frecuentes** (que aparecen en muy pocos documentos) o **demasiado comunes** (que aparecen en casi todos los documentos), según umbrales definidos.

Estas funciones permiten mejorar el enfoque analítico y la eficiencia computacional en tareas posteriores como análisis de sentimientos o modelamiento de tópicos.


In [16]:
?dfm_select()

dfm_select              package:quanteda               R Documentation

_S_e_l_e_c_t _f_e_a_t_u_r_e_s _f_r_o_m _a _d_f_m _o_r _f_c_m

_D_e_s_c_r_i_p_t_i_o_n:

     This function selects or removes features from a dfm or fcm, based
     on feature name matches with ‘pattern’.  The most common usages
     are to eliminate features from a dfm already constructed, such as
     stopwords, or to select only terms of interest from a dictionary.

_U_s_a_g_e:

     dfm_select(
       x,
       pattern = NULL,
       selection = c("keep", "remove"),
       valuetype = c("glob", "regex", "fixed"),
       case_insensitive = TRUE,
       min_nchar = NULL,
       max_nchar = NULL,
       padding = FALSE,
       verbose = quanteda_options("verbose")
     )
     
     dfm_remove(x, ...)
     
     dfm_keep(x, ...)
     
     fcm_select(
       x,
       pattern = NULL,
       selection = c("keep", "remove"),
       valuetype = c("glob", "regex", "fixed"),
   

#### **`dfm_remove`()**

Supongamos que no eliminamos los stop-words cuando tokenizamos el documento.

In [17]:
toks <- tokens(data_corpus_inaugural)


In [18]:
dfm_inaugural_nostop <- dfm(tokens(data_corpus_inaugural)) %>%
             dfm_remove(pattern = stopwords("en"))

print(dfm_inaugural_nostop)

Document-feature matrix of: 60 documents, 9,453 features (92.73% sparse) and 4 docvars.
                 features
docs              fellow-citizens senate house representatives : among
  1789-Washington               1      1     2               2 1     1
  1793-Washington               0      0     0               0 1     0
  1797-Adams                    3      1     0               2 0     4
  1801-Jefferson                2      0     0               0 1     1
  1805-Jefferson                0      0     0               0 0     7
  1809-Madison                  1      0     0               0 0     0
                 features
docs              vicissitudes incident life event
  1789-Washington            1        1    1     2
  1793-Washington            0        0    0     0
  1797-Adams                 0        0    2     0
  1801-Jefferson             0        0    1     0
  1805-Jefferson             0        0    2     0
  1809-Madison               0        0    1     0
[ reac

#### **`dfm_trim()`**

Los argumentos de esta función *min_termfreq* y *max_termfreq* permiten determinar el umbral de frecuencia para el análisis.

In [19]:
dfm_trim(dfm_inaugural_nostop , min_termfreq = 20)


Document-feature matrix of: 60 documents, 675 features (59.91% sparse) and 4 docvars.
                 features
docs              fellow-citizens : among life greater order   , day present  .
  1789-Washington               1 1     1    1       1     2  70   2       5 23
  1793-Washington               0 1     0    0       0     0   5   0       1  4
  1797-Adams                    3 0     4    2       0     4 201   1       2 33
  1801-Jefferson                2 1     1    1       1     1 128   1       0 37
  1805-Jefferson                0 0     7    2       0     3 142   1       3 41
  1809-Madison                  1 0     0    1       0     0  47   0       1 21
[ reached max_ndoc ... 54 more documents, reached max_nfeat ... 665 more features ]

#### **`dfm_select()`**

¿Cuántas palabras "largas" hay en los documentos del corpus?


In [20]:
dfmat_inaug_long <- dfm_select(dfm_inaugural_nostop, min_nchar = 10)
print(dfmat_inaug_long)

Document-feature matrix of: 60 documents, 2,472 features (95.12% sparse) and 4 docvars.
                 features
docs              fellow-citizens representatives vicissitudes notification
  1789-Washington               1               2            1            1
  1793-Washington               0               0            0            0
  1797-Adams                    3               2            0            0
  1801-Jefferson                2               0            0            0
  1805-Jefferson                0               0            0            0
  1809-Madison                  1               0            0            0
                 features
docs              transmitted veneration predilection flattering inclination
  1789-Washington           1          1            1          1           1
  1793-Washington           0          0            0          0           0
  1797-Adams                0          2            0          0           1
  1801-Jefferson    

Quanteda es treendamente flexible y permite combinar estos comandos.

In [21]:
dfm_inaug__stem_docfreq <- dfm_trim(dfm_inaug_stem, min_docfreq = 0.90, docfreq_type = "prop")

print(dfm_inaug__stem_docfreq)

Document-feature matrix of: 60 documents, 14 features (6.90% sparse) and 4 docvars.
                 features
docs              countri can time citizen nation may peopl govern great now
  1789-Washington       5   9    1       4      4   6     4      9     3   1
  1793-Washington       1   0    0       1      0   1     1      1     0   2
  1797-Adams           10   9    3       3     24  13    20     20     5   0
  1801-Jefferson        4   3    0       5      4   8     2     14     2   1
  1805-Jefferson        5   6    6      10      6  10     0      3     1   3
  1809-Madison          5   5    1       0     10   1     1      0     0   1
[ reached max_ndoc ... 54 more documents, reached max_nfeat ... 4 more features ]


Sólo 11 características aparecen en más del 90% de los documentos....pero sólo vemos 10!

In [22]:
print(topfeatures(dfm_inaug_stem,11, scheme = "docfreq"))


 nation   peopl countri     can   great      us   right    time  govern   world 
     59      58      57      57      57      57      56      55      55      55 
citizen 
     54 
